File: path_sum_II_backtracking.java
// Time Complexity :O(n), where n is the number of node in the tree, (time required to traverse the tree)
// Space Complexity :O(n), where n is the height of the tree (worst case a skwed binary tree)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

//Three liner explanation of your code in plain english
//1. we can use inorder traversal to traverse the complete tree. At every node calculate the runningSum and register the path till root
        // node in a single array
//2. If the root is leaf node and the sum is equal to the runningSum, then add the path of the root as a new arrayList in the result
//3. Keep exploring the left and the right subtrees for the complete tree, to find multiple path if there

// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //List to return result
    List<List<Integer>> result;
    int target;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        result = new ArrayList<>();
        if(root == null) return result;
        target = sum;
        helper(root, 0, new ArrayList<>());
        return result;
    }
    
    private void helper( TreeNode root, int runningSum, List<Integer> path){
        //base
        if(root == null) return;
        
        //logic
        //ACTION
        //register the path till the current node
        path.add(root.val);
        //calculate the runningSum till the current node
        runningSum += root.val;
        
        //add the path to the result if get the sum and the current node id leaf
        if( root.left ==null && root.right ==null && runningSum == target){
            result.add(new ArrayList<>(path));
        }
        //RECURRSION
        //traverse the left subtree with new instance of the path
        helper(root.left, runningSum, path);
        //traverse the right subtree with new instance of the path
        helper(root.right, runningSum, path);
        
        //BACK TRACKING
        //undo last action (remove the last added root)
        path.remove(path.size()-1);
    }
}

File: PathSum_II_Backtracking.py
# Created by Aashish Adhikari at 1:18 PM 1/15/2021

'''
Time Complexity:
Ignoring the deepcopy operation when a viable path is found, O(n).
Considering those operations, in the worst case, all the leaves with give a viable path. Since there will be a max of O( 2 ^ h)  leaves, TC :  O(n + 2 ^ h)

Space Complexity:
O(n) since in the worst case, we will add all elements to the array variable that stores the current path.
'''


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right



class Solution(object):

    def backtrack(self, node, current_path_upto_parent, sum_upto_parent):



        # base case
        current_path_upto_parent.append(node.val)
        sum_upto_parent += node.val
        if node.left is None and node.right is None:

            if sum_upto_parent == self.sum:
                self.paths.append(copy.deepcopy(current_path_upto_parent))



        # recursion
        if node.left is not None:
            self.backtrack(node.left, current_path_upto_parent, sum_upto_parent)     #<-- this traversal does not affect the state of the current path when it returns back

        if node.right is not None:
            self.backtrack(node.right, current_path_upto_parent, sum_upto_parent)    #<-- this traversal does not affect the state of the current path when it returns back

        # backtracking
        # once recursion returns from both the children, remove the current node element from the path since it returns the recursion to its parent
        current_path_upto_parent.pop()

    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """

        if root is None:
            return []

        self.sum = sum
        self.paths = []

        self.backtrack(root, [], 0)

        return self.paths